### ISSUE TITLE
Anchor Creation from Feature Map

### ISSUE THREAD

**Issue by brian-regan** [mmr12/DeepBlueSea#3]
Given the feature map outputs of the RPN we want to actually make the proposals which are then passed onto a classifier.

**Comment by brian-regan** [comment]
operating on branch _anchors_

**Comment by brian-regan** [comment]
See branch [here](https://github.com/mmr12/DeepBlueSea/tree/anchors) and _early stage_ FasterRCNN model [here](https://github.com/mmr12/DeepBlueSea/blob/anchors/models/faster_rcnn_model.py)

**Comment by brian-regan** [comment]
TODO: write a data loader

**Comment by brian-regan** [comment]
Have the bones done out, except for things pertaining to loss (see [here](https://github.com/mmr12/DeepBlueSea/commit/3fb4f4201d58fa2c5845fd6742e586c974a3255f)). Am currently working on loading in and formatting the labels and calculating a loss.

**Comment by brian-regan** [comment]
Fixed the data loading problem. Had to make think in a slightly different way than to the skeleton in that we use a tensorflow dataset object as a loader rather than passing raw numpy files into the placeholders through a feed dict. Using the tensorflow dataset pipeline seems to be the way to go. Thus you only pass a "handle" in the feed dict and then that refers to a dataset iterator defined [here](https://github.com/mmr12/DeepBlueSea/blob/anchors/data_loader/faster_rcnn_loader.py)

**Comment by brian-regan** [comment]
Also added a summary ability for images and things that are parts of the model (rather than say loss which would be an output). Couldnt use the logger wrapper they use because it requires non tensor inputs so id have to convert the images to numpy arrays just to summarise them which re-tensorifies them (see: [here](https://github.com/mmr12/DeepBlueSea/blob/anchors/utils/logger.py)). 
However, I use their self.train_summary_writer, and just insert a new line in the [training file](https://github.com/mmr12/DeepBlueSea/blob/anchors/trainers/faster_rcnn_trainer.py) that logs all summaries created within the model

**Comment by brian-regan** [comment]
Left to do atm: 

From a csv file which @mmr12 created, fill in the y matrix (currently two 768, 768, 2*n_boxes and 768, 768, 4*n_boxes arrays so as to match output)  with the appropriate values. This is tedious enough but not "hard". Will start asap. 

Write a loss function for when you have the real and outputed versions of these y matrices

**Comment by brian-regan** [comment]
Moved the y unwrapping to the model size. Runs quickly it seems on GPU. Implemented loss. Note, made a small change to the output of the classification branch. It now outputs a tensor with depth n_anchors rather than 2*n_anchors since its easier to calculate loss then with [tf.nn.sigmoid_cross_entropy_with_logits](https://www.tensorflow.org/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits) since the y labels also have this depth. It is equivalent anyways since if you had a 2*n_anchor depth you would just normalise the probabilities first anyways

**Comment by mmr12** [comment]
See notebook "anchor_code" for test of anchors

observations:
at the moment, we do:
for each anchor shape:
calculate IOU for each boat
keep the largest IOU
threshold that IOU (to 0.5), keep larger vals

note that in the paper they take 2 additional steps:
1) keep the anchor with highest IOU [unclear what "highest" is calculated with]
2) assign a negative label to anchors with IOU < 30%
3) discard anchors in the middle

also notice that for the example in the notebook, the max IOU is 17%, which probably means we need to change anchor sizes

**Comment by brian-regan** [comment]
> * keep the anchor with highest IOU [unclear what "highest" is calculated with]

What do you mean exactly? In the Faster RCNN paper they take all IOU's above .7 and in the rare case that there are none above 0.7 take the top 100 or so.
